compiler/front_end/synthetics.py

```diff
@@ -215,23 +216,118 @@ def _add_size_virtuals(structure, type_definition):
   structure.field.extend([size_field])
 
 
+# The replacement for the "$next" keyword is a simple "start + size" expression.
+# 'x' and 'y' are placeholders, to be replaced.
+_NEXT_KEYWORD_REPLACEMENT_EXPRESSION = expression_parser.parse("x + y")
+
+
+def _maybe_replace_next_keyword_in_expression(expression, last_location,
+                                              source_file_name, errors):
+  if not expression.HasField("builtin_reference"):
+    return
+  if expression.builtin_reference.canonical_name.object_path[0] != "$next":
+    return
+  if not last_location:
+    errors.append([
+      error.error(source_file_name, expression.source_location,
+                  "`$next` may not be used in the first physical field of a " +
+                  "structure; perhaps you meant `0`?")
+    ])
+    return
+  original_location = expression.source_location
+  expression.CopyFrom(_NEXT_KEYWORD_REPLACEMENT_EXPRESSION)
+  expression.function.args[0].CopyFrom(last_location.start)
+  expression.function.args[1].CopyFrom(last_location.size)
+  expression.source_location.CopyFrom(original_location)
+  # DO NOT SUBMIT: Error messages for circular references?
```

Comment:
Anything left to do for this `DO NOT SUBMIT`?

Reply:
Yes... the message is a bit less clear than I would prefer, but dependency cycles that include `$next` are correctly flagged as errors.